zanata-war/src/main/java/org/zanata/dao/TextFlowDAO.java

```diff
@@ -300,4 +302,73 @@ public long countActiveTextFlowsInProjectIteration(Long versionId) {
         return builder.build();
     }
 
+    /**
+     * Return text flows from HProjectIteration#id=sourceVersionId and
+     * HProjectIteration#id=targetVersionId which has matching content
+     * hash and document id.
+     *
+     * @param sourceVersionId
+     * @param targetVersionId
+     * @param offset
+     * @param maxResults
+     */
+    public List<HTextFlow[]> getSourceByMatchedContext(Long sourceVersionId,
+        Long targetVersionId, int offset, int maxResults) {
+        StringBuilder queryBuilder = new StringBuilder();
+        queryBuilder
+            .append("select fromTF, toTF from HTextFlow fromTF, HTextFlow toTF ")
+            .append("where fromTF.document.projectIteration.id = :sourceVersionId ")
+            .append("and toTF.document.projectIteration.id = :targetVersionId ")
+            .append("and fromTF.obsolete = false ")
+            .append("and fromTF.document.obsolete = false ")
+            .append("and toTF.obsolete = false ")
+            .append("and toTF.document.obsolete = false ")
+            .append("and fromTF <> toTF ")
+            .append("and fromTF.contentHash = toTF.contentHash ")
+            .append("and fromTF.document.docId = toTF.document.docId");
+
+        Query query = getSession()
+            .createQuery(queryBuilder.toString())
+            .setParameter("sourceVersionId", sourceVersionId)
+            .setParameter("targetVersionId" , targetVersionId)
+            .setMaxResults(maxResults)
+            .setFirstResult(offset)
+            .setCacheable(true)
+            .setComment("TextFlowDAO.getTranslationsByMatchedContext");
+
+        List<HTextFlow[]> results = Lists.newArrayList();
+        for(Object result: query.list()) {
+            Object[] castResults = (Object[]) result;
+            results.add(new HTextFlow[] { (HTextFlow) castResults[0],
+                (HTextFlow) castResults[1] });
+        }
+        return results;
+    }
+
+    public int getSourceByMatchedContextCount(Long sourceVersionId,
+        Long targetVersionId) {
```

Comment:
This is mostly the same as `getSourceByMatchedContext`, just the first line of the query, a few params, and the way that results are collated look different. Could a method be added that takes the `StringBuilder` after the select clause is set, and add all the common lines to it?